topology/src/main/java/io/camunda/zeebe/topology/util/TopologyUtil.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.topology.util;
+
+import io.atomix.cluster.MemberId;
+import io.atomix.primitive.partition.PartitionId;
+import io.atomix.primitive.partition.PartitionMetadata;
+import io.camunda.zeebe.topology.state.ClusterChangePlan;
+import io.camunda.zeebe.topology.state.ClusterTopology;
+import io.camunda.zeebe.topology.state.MemberState;
+import io.camunda.zeebe.topology.state.PartitionState;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public final class TopologyUtil {
+
+  private TopologyUtil() {}
+
+  public static ClusterTopology getClusterTopologyFrom(
+      final Set<PartitionMetadata> partitionDistribution) {
+    final var partitionsOwnedByMembers =
+        partitionDistribution.stream()
+            .flatMap(
+                p ->
+                    p.members().stream()
+                        .map(m -> Map.entry(m, Map.entry(p.id().id(), p.getPriority(m)))))
+            .collect(
+                Collectors.groupingBy(
+                    Entry::getKey,
+                    Collectors.toMap(
+                        e -> e.getValue().getKey(),
+                        e -> PartitionState.active(e.getValue().getValue()))));
+
+    final var memberStates =
+        partitionsOwnedByMembers.entrySet().stream()
+            .collect(
+                Collectors.toMap(Entry::getKey, e -> MemberState.initializeAsActive(e.getValue())));
+
+    return new io.camunda.zeebe.topology.state.ClusterTopology(
+        0, memberStates, ClusterChangePlan.empty());
+  }
```

Comment:
:thought_balloon: took me a moment to understand what's going on here so I tried to rewrite it (without using streams even). I'm not sure if that's really better though.

```java
  public static ClusterTopology getClusterTopologyFrom(
      final Set<PartitionMetadata> partitionDistribution) {

    final var partitionStatesByMember = new HashMap<MemberId, Map<Integer, PartitionState>>();
    for (final var partitionMetadata : partitionDistribution) {
      final var partitionId = partitionMetadata.id().id();
      for (final var member : partitionMetadata.members()) {
        final var memberPriority = partitionMetadata.getPriority(member);
        partitionStatesByMember
            .computeIfAbsent(member, k -> new HashMap<>())
            .put(partitionId, PartitionState.active(memberPriority));
      }
    }
    final var memberStates = new HashMap<MemberId, MemberState>();
    for (final var e : partitionStatesByMember.entrySet()) {
      memberStates.put(e.getKey(), MemberState.initializeAsActive(e.getValue()));
    }

    return new io.camunda.zeebe.topology.state.ClusterTopology(
        0, memberStates, ClusterChangePlan.empty());
  }
```

Reply:
It would be nice if we don't have to convert between `PartitionMetadata` and `ClusterTopology` and just use one representation for both. Not sure if that is feasible though. Anyway don't want to do that now. 

I will check if I can improve the readability here. May be not using streams is better.